manager/allocator/network/allocator.go

```diff
@@ -0,0 +1,869 @@
+	// for allocating ports
+	//
+	// allocate ports first. calling port.Allocate doesn't actually alter the
+	// state of the port allocator, it just returns a Proposal or an error. If
+	// there is an error then port allocation is not possible, and we can stop
+	// here. Otherwise, the returned proposal contains the information which is
+	// needed to alter the state of the port allocator and commit the changes.
+	// If a proposal is returned, then committing the proposal is guaranteed to
+	// succeed.
+	//
+	// When the code in this function is unpacked, it looks (essentially) like
+	// this:
+	//
+	//   Parts that can fail but be undone:
+	//     port.Allocate: Allocate Ports (verifying if allocation _will_ succeed)
+	//     ipam.AllocateVIPs: Allocate new VIPs
+	//   Parts that cannot fail but cannot be undone:
+	//     in the same call to ipam.AllocateVIPs: Deallocate VIPs no longer in use
+	//     proposal.Commit: Commit the port changes (which we know to be valid)
+	//
+	// This complication does not exist in networks or tasks, which are only
+	// one-way, and therefore only acquire resources or release them, never
+	// both at the same time like services, and it does not exist on nodes
+	// because nodes only worry about one resource type (attachments).
+	proposal, err := a.port.Allocate(endpoint, service.Spec.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	// TODO(dperny) this handles the case of spec.Networks, which we should
+	// deprecate before removing this code entirely
+	networks := service.Spec.Task.Networks
+	if len(service.Spec.Task.Networks) == 0 && len(service.Spec.Networks) != 0 {
+		networks = service.Spec.Networks
+	}
+	ids := make(map[string]struct{}, len(networks))
+	// build up a list of network ids to allocate vips for
+	for _, nw := range networks {
+		// we don't allocate VIPs for node-local networks, so if the network is
+		// found in the nodeLocalNetworks set, skip it
+		if _, ok := a.nodeLocalNetworks[nw.Target]; !ok {
+			ids[nw.Target] = struct{}{}
+		}
+	}
+
+	// ingress is special because it cannot be normally attached to and so will
+	// not be found in the spec's NetworkAttachmentConfigs. however, in the
+	// actual objects, it should have a VIP. so, if we need it, append it to
+	// the list of network IDs we're requesting VIPs for.
+	if ingressNeeded(proposal.Ports()) {
+		if a.ingressID != "" {
+			ids[a.ingressID] = struct{}{}
+		} else {
+			// if we need an ingress network, but we don't currently have one,
+			// then the dependencies for this service are not allocated. if we
+			// didn't return an error here, we would still get an
+			// ErrDependencyNotAllocated, but it would be from the IPAM
+			// allocator, complaining that network "" (empty string) was not
+			// allocated.
+			return errors.ErrDependencyNotAllocated("network", "ingress")
+		}
+	}
+
+	// check the resolution mode of the new endpoint spec. If it's
+	// ResolutionModeDNSRoundRobin, then we need to deallocate vips, not
```

Comment:
`ResolutionModeDNSRoundRobin` means that service discovery, instead of returning a VIP, returns a list of IP addresses (or a single IP address, different each time. something like that) for all of the tasks.

Reply:
Got it, thanks.  Could that explanation be added to the comment?